File: xamarin/Xtrade/Xtrade.AndroidApp/Adapters/RatesRecyclerAdapter.cs
namespace Xtrade.AndroidApp.Adapters
{
    using System;
    using System.Collections.Generic;
    using Android.Support.V7.App;
    using Android.Support.V7.Widget;
    using Android.Views;
    using Shared.Interfaces.Domain.Models;
    using ViewHolders;

    public class RatesRecyclerAdapter : RecyclerView.Adapter
    {
        private IList<IRate> _allRates;
        private readonly AppCompatActivity _activity;
        private readonly Action<int> _itemClickAction;

        public RatesRecyclerAdapter(AppCompatActivity activity, IList<IRate> allRates, Action<int> itemClickAction)
        {
            this._allRates = allRates;
            this._activity = activity;
            this._itemClickAction = itemClickAction;
        }

        public void UpdateDataSet(IList<IRate> allRates)
        {
            this._allRates = allRates;
            this.NotifyDataSetChanged();
        }

        public override int ItemCount
        {
            get { return this._allRates != null ? this._allRates.Count : 0; }
        }

        public IRate this[int position]
        {
            get { return this._allRates[position]; }
        }

        public override RecyclerView.ViewHolder OnCreateViewHolder(ViewGroup parent, int viewType)
        {
            View itemView = LayoutInflater.From(parent.Context).Inflate(Resource.Layout.recycler_item_rate, parent, false);
            return new RateViewHolder(itemView, this._itemClickAction);
        }

        public override void OnBindViewHolder(RecyclerView.ViewHolder holder, int position)
        {
            RateViewHolder viewHolder = holder as RateViewHolder;
            IRate rate = this._allRates != null ? this._allRates[position] : null;

            if (rate != null && viewHolder != null)
            {
                int flagID = this._activity.Resources.GetIdentifier("flag_" + rate.CurrencyCode.ToLower(), "drawable", this._activity.PackageName);

                if (flagID == 0)
                {
                    viewHolder.ForexFlagImageView.Visibility = ViewStates.Gone;
                }
                else
                {
                    viewHolder.ForexFlagImageView.Visibility = ViewStates.Visible;
                    viewHolder.ForexFlagImageView.SetImageDrawable(this._activity.Resources.GetDrawable(flagID));
                }

                viewHolder.CountryNameTextView.Text = rate.Description;
                viewHolder.ForexCodeTextView.Text = rate.CurrencyCode;
                viewHolder.Rate1TextView.Text = rate.BuysNotes.ToString("C0");
                viewHolder.Rate2TextView.Text = rate.BuysCheques.ToString("C0");
                viewHolder.Rate3TextView.Text = rate.SellsNotes.ToString("C0");
            }
        }
    }
}

File: android/Xtrade/app/src/main/java/nz/co/fendustries/xtrade/adapters/RatesRecyclerAdapter.java
package nz.co.fendustries.xtrade.adapters;

import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;
import java.util.Locale;

import nz.co.fendustries.xtrade.R;
import nz.co.fendustries.xtrade.adapters.viewHolders.RateViewHolder;
import nz.co.fendustries.xtrade.domain.models.Rate;
import nz.co.fendustries.xtrade.interfaces.RecyclerViewItemTapCallback;

/**
 * Created by joshuafenemore on 21/10/16.
 */
public class RatesRecyclerAdapter extends RecyclerView.Adapter<RateViewHolder>
{
    private java.text.NumberFormat currencyInstance = java.text.NumberFormat.getCurrencyInstance(Locale.US);

    private List<Rate> allRates;
    private AppCompatActivity activity;
    private RecyclerViewItemTapCallback recyclerViewItemTapCallback;

    public RatesRecyclerAdapter(AppCompatActivity activity, List<Rate> allRates, RecyclerViewItemTapCallback recyclerViewItemTapCallback)
    {
        this.allRates = allRates;
        this.activity = activity;
        this.recyclerViewItemTapCallback = recyclerViewItemTapCallback;
    }

    @Override
    public int getItemCount()
    {
        return this.allRates != null ? this.allRates.size() : 0;
    }


    @Override
    public RateViewHolder onCreateViewHolder(ViewGroup parent, int viewType)
    {
        View itemView = this.activity.getLayoutInflater().inflate(R.layout.recycler_item_rate, parent, false);
        return new RateViewHolder(itemView, this.recyclerViewItemTapCallback);
    }

    @Override
    public void onBindViewHolder(final RateViewHolder viewHolder, int position)
    {
        Rate rate = this.allRates != null ? this.allRates.get(position) : null;

        if (rate != null && viewHolder != null)
        {
            int flagID = this.activity.getResources().getIdentifier("flag_" + rate.getCurrencyCode().toLowerCase(), "drawable", this.activity.getPackageName());

            if (flagID == 0)
            {
                viewHolder.getForexFlagImageView().setVisibility(View.GONE);
            }
            else
            {
                viewHolder.getForexFlagImageView().setVisibility(View.VISIBLE);
                viewHolder.getForexFlagImageView().setImageDrawable(this.activity.getResources().getDrawable(flagID));
            }

            viewHolder.getForexCodeTextView().setText(rate.getCurrencyCode());
            viewHolder.getBuyRateTextView().setText(currencyInstance.format(rate.getBuysNotes()));
            viewHolder.getSellRateTextView().setText(currencyInstance.format(rate.getSellsNotes()));
        }
    }
}